Make profiler backend modular, add support for vernier profiler (#101)
Shopify/app_profiler: README.md

| Key | Value | Notes |
| --- | ----- | ----- |

> | profile/mode | Supported profiling modes: `cpu`, `wall`, `object` for stackprof. | Use `profile` in (1), and `mode` in (2). Vernier backend only supports `wall` and `retained` at present time|

Reviewer: ```suggestion
| profile/mode | Supported profiling modes: `cpu`, `wall`, `object` for stackprof. | Use `profile` in (1), and `mode` in (2). Vernier backend only supports `wall` and `retained` at present time. |
```

| async | Upload profile in a background thread. When this is set, profile redirect headers are not present in the response.
| interval | Sampling interval in microseconds. | |
| ignore_gc | Ignore garbage collection frames | |
| autoredirect | Redirect request automatically to Speedscope's page after profiling. | |
| context | Directory within the specified bucket in the selected storage where raw profile data should be written. | Only supported in (2). Defaults to `Rails.env` if not specified. |

> | backend | Profiler to use, either `stackprof` or `vernier`. Defaults to `stackprof`. Note that Vernier requires Ruby 3.2.1+ |

Reviewer: ```suggestion
| backend | Profiler to use, either `stackprof` or `vernier`. Defaults to `stackprof`. Note that Vernier requires Ruby 3.2.1+. |
```

> | backend | Profiler to use, either `stackprof` or `vernier`. Defaults to stackprof. Vernier is only available when added separately in the application Gemfile, and the Ruby version is at least 3.2.1 |

Reviewer: ```suggestion
| backend | Profiler to use, either `stackprof` or `vernier`. Defaults to `stackprof`. Note that Vernier requires Ruby 3.2.1+ |
```
Do we expect apps to use multiple backends? If not we could just use the accessor configs to be one or the other.

Rails.application.config.app_profiler.middleware_action = AppProfiler::Middleware::UploadAction
```

## Profiler backends

> It is possible to configure app_profiler to use the [`vernier`](https://github.com/jhawthorn/vernier) or stackprof. To use vernier, it must be added separately in the application Gemfile.

Reviewer: ```suggestion
It is possible to configure AppProfiler to use the [`vernier`](https://github.com/jhawthorn/vernier) or [`stackprof`](https://github.com/tmm1/stackprof). To use `vernier`, it must be added separately in the application Gemfile.
```

I wonder if we should be making stackprof a peer dependency as well (not in the gemspec, but required if defined in the bundle like vernier is). Maybe we can do that in a followup.

The backend can be selected dynamically at runtime using the `backend` parameter. The default backend to use when this parameter is not specified can be configured with:

```ruby

> AppProfiler.profiler_backend = AppProfiler::StackprofBackend # or AppProfiler::VernierBackend

Author: TODO update to backend=
